File: Leetcode 102 - binary-tree-level-order-traversal/102.binary-tree-level-order-traversal.cpp
/*
 * @lc app=leetcode id=102 lang=cpp
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  vector<vector<int>> levelOrder(TreeNode *root) {
    vector<vector<int>> tree;

    if (root == NULL)
      return tree;

    queue<TreeNode *> q;

    q.push(root);

    while (!q.empty()) {
      int len = q.size();
      vector<int> curTree;

      for (int i = 0; i < len; i++) {
        TreeNode *cur = q.front();
        q.pop();

        if (cur->left)
          q.push(cur->left);
        if (cur->right)
          q.push(cur->right);

        curTree.push_back(cur->val);
      }

      tree.push_back(curTree);
    }

    return tree;
  }
};
// @lc code=end

File: Leetcode 102 - binary-tree-level-order-traversal/102.binary-tree-level-order-traversal.java
/*
 * @lc app=leetcode id=102 lang=java
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  public List<List<Integer>> levelOrder(TreeNode root) {
    List<List<Integer>> tree = new ArrayList<>();

    if (root == null) return tree;

    Queue<TreeNode> q = new LinkedList<>();

    q.offer(root);

    while (!q.isEmpty()) {
      List<Integer> curTree = new ArrayList<>();
      int len = q.size();

      while (len > 0) {
        TreeNode cur = q.poll();

        if (cur != null) {
          if (cur.left != null) {
            q.offer(cur.left);
          }
          if (cur.right != null) {
            q.offer(cur.right);
          }
          curTree.add(cur.val);
        }

        len--;
      }

      if (curTree.size() > 0) {
        tree.add(curTree);
      }
    }

    return tree;
  }
}
// @lc code=end
